orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

```diff
@@ -654,36 +654,63 @@ def test_changes_input_requires_grad_reinitializes_module_gradient_graph_builder
 @pytest.mark.parametrize("device", ['cuda'])
 def test_input_requires_grad_backward_creates_input_grad_as_required0(device):
     N, D_in, H, D_out = 32, 784, 500, 10
-    model = NeuralNetMultiplePositionalArgumentsMultipleOutputs0(D_in, H, D_out).to(device)
-    model = ORTModule(model)
-    x1 = torch.randn(N, D_in, device=device, requires_grad=True)
-    x2 = torch.randn(N, D_in, device=device, requires_grad=True)
-
-    y1, _ = model(x1, x2)
-    s1 = y1.sum()
-    s1.backward()
-    assert x1.grad is not None and x2.grad is not None
-
-    # named_params[0] and named_params[1] correspond to weight and bias for fc1, similarly
-    # named_params[2] and named_params[3] correspond to weight and bias for fc2.
-    named_params = list(model.named_parameters())
-    assert torch.count_nonzero(named_params[0][1].grad) > 0
-    assert torch.count_nonzero(named_params[1][1].grad) > 0
-    assert named_params[2][1].grad is None or torch.count_nonzero(named_params[2][1].grad) == 0
-    assert named_params[3][1].grad is None or torch.count_nonzero(named_params[3][1].grad) == 0
-
-    # Reset gradients
-    for param in named_params:
-        param[1].grad = None
-
-    _, y2 = model(x1,x2)
-    s2 = y2.sum()
-    s2.backward()
-    named_params = list(model.named_parameters())
-    assert named_params[0][1].grad is None or torch.count_nonzero(named_params[0][1].grad) == 0
-    assert named_params[1][1].grad is None or torch.count_nonzero(named_params[1][1].grad) == 0
-    assert torch.count_nonzero(named_params[2][1].grad) > 0
-    assert torch.count_nonzero(named_params[3][1].grad) > 0
+    pt_model = NeuralNetMultiplePositionalArgumentsMultiOutputsWithoutDependency(D_in, H, D_out).to(device)
+    ort_model = ORTModule(copy.deepcopy(pt_model))
+    pt_x1 = torch.randn(N, D_in, device=device, requires_grad=True)
+    pt_x2 = torch.randn(N, D_in, device=device, requires_grad=True)
+    ort_x1 = pt_x1.clone().detach()
+    ort_x2 = pt_x2.clone().detach()
+    ort_x1.requires_grad = True
+    ort_x2.requires_grad = True
+
+    def run_step0(model, x1, x2):
+        y1, _ = model(x1, x2)
+        s1 = y1.sum()
+        s1.backward()   # y2's gradient will be materialized to full shape.
+
+    run_step0(pt_model, pt_x1, pt_x2)
+    run_step0(ort_model, ort_x1, ort_x2)
+
+    assert torch.allclose(ort_x1.grad, pt_x1.grad)
```

Comment:
let's also assert on the forward results? 
I caught a mismatch on forward result in test_input_requires_grad_backward_creates_input_grad_as_required1
see 
https://msdata.visualstudio.com/Vienna/_sprints/taskboard/ONNX%20Training/Vienna/Cobalt?workitem=1064208